### ISSUE TITLE
Inefficiencies with shouldRender

### ISSUE THREAD

**Issue by mplis-jetsetter** [mozilla-services/react-jsonschema-form#255]
### Prerequisites

- [x] I have read the [documentation](https://github.com/mozilla-services/react-jsonschema-form/blob/master/README.md#readme).

### Description

I believe the `shouldRender` function in `utils.js` is returning `true` in too many instances, particularly for nested `ArrayField`s or `ObjectField`s, because it's unnecessarily requiring function equality in props and state.

### Steps to Reproduce

1. In your clone of this repository, add this line to `shouldRender`: `console.log(comp.props, nextProps, !deeper(comp.props, nextProps) || !deeper(comp.state, nextState));`
2. Start the development server with `npm start`
3. Go to `localhost:8080` and click on the "Nested" tab
4. Start typing in the "Task list title" field
5. Look in the console logs for something like `Object {name: "tasks"...`. This is where `shouldRender` was called for the "tasks" `ArrayField`
6. Observe that the objects logged on that line are structurally equivalent, but that `shouldRender` is still returning `true`. I believe that this is happening only because the `onChange` functions in those objects are referentially non-equal.

#### Expected behavior

I would expect that `shouldRender` for the "tasks" `ArrayField` would return `false` and that the component does not re-render, because none of the data that affects how the component renders has changed. Re-rendering all `ArrayField`s or `ObjectField`s when none of the data in that component has changed, particularly in large forms, is a performance drain.

#### Actual behavior

`ArrayField` re-renders when making a change to another part of the form.

### Version

0.34.1

### Proposed Solution

I think that a function like the one below should be added to `utils.js` and called on `comp.props` and `nextProps` before they're passed to `deeper`

```
function _removeFunctionsFromObj(obj) {
  var newObj = {};
  for (var key in obj) {
    if (typeof obj[key] !== 'function') {
      newObj[key] = obj[key];
    }
  }
  return newObj;
}
```

I'd be happy to submit a PR, but I thought it might be better to start a discussion first in case there's another way to solve this problem.

**Comment by n1k0** [comment]
This is exactly why I used instance properties for many methods in the components API; the root of the issue is binding:

```js
> function foo(){}
undefined
> foo === foo
true
> foo.bind(null) === foo
false
```

Instance function properties being pre-bound, they're always passed down to children components as a static reference, so they are comparable.

In theory we shouldn't have so many render calls, maybe we forgot to use instance properties in some place in the code?

As to removing all functions from props objects before comparing them, I'm a little nervous with this idea as it may be important to trigger a rerender if the function has actually changed and requires a rerender.

Feel free to send a PR with the latter approach so we can stare at the test suite and play with the playground to figure out if that works.

**Comment by n1k0** [comment]
Also to be noted that the playground is a very special example as the data flow is bi-directional: every time you enter a value eThat's tricky in a field, it updates the state JSON editor, and every time you update the state or form props JSON editors these will be passed as updated props to the Form object, triggering necessary renders. Most often in real world apps this shouldn't happen much.

**Comment by mplis-jetsetter** [comment]
> This is exactly why I used instance properties (or class fields) for many methods in the components API; the root of the issue is binding

I'm not super familiar with ES6 classes yet, but is this related to why the Form's [`onChange`](https://github.com/mplis-jetsetter/react-jsonschema-form/blob/master/src/components/Form.js#L78) method is defined like this:

     onChange = (...) => {...}

instead of this?:

    onChange(...) {...}

> As to removing all functions from props objects before comparing them, I'm a little nervous with this idea as it may be important to trigger a rerender if the function has actually changed and requires a rerender.

I completely understand your concern. I think it'd be preferable if we could figure out why a new reference to `onChange` is being passed in with `nextProps`.

> Also to be noted that the playground is a very special example as the data flow is bi-directional

The playground seemed like the easiest way for me to show how to reproduce the issue I was looking at. I'm definitely seeing the same behavior in my own code. I made a change similar to the one I mentioned above in my own code and I'm seeing some noticeable performance benefits.

**Comment by n1k0** [comment]
> I'm not super familiar with ES6 classes yet, but is this related to why the Form's `onChange` method is defined like this

Yes. What you want to avoid is:

```jsx
render() {
  return <Child onChange={this.onChange.bind(this)} />;
}
```

Because that means mutating the reference to the function at each render, so the `onChange` would always be detected as different than previously:

```js
> foo.bind(null) === foo.bind(null)
false
> a = {}
Object {}
> foo.bind(a) === foo.bind(a)
false
```

With a class field, the method is already bound to the current instance so it's the very same function:

```js
class Foo {
  onChange = () => {}
}
const foo = new Foo();
foo.onChange === foo.onChange // true
```

**Comment by n1k0** [comment]
One last thing, we already have [unit tests](https://github.com/mozilla-services/react-jsonschema-form/blob/9f984590df1e73e1be3ff2cd3da513c961f5c698/test/performance_test.js) ensuring that we don't rerender when props for `Form`, `ArrayField` and `ObjectField` see new equivalent props passed; maybe adding more tests to this suite is a good way to figure out actual problems if any.

**Comment by n1k0** [comment]
Released in [v0.36.0](https://github.com/mozilla-services/react-jsonschema-form/releases/tag/v0.36.0).
